### ISSUE TITLE
Add a MustHappen() method to define expectations

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#464]
Currently, if you want to define the behavior for a method on a fake and check that the method was called, you need to call `A.CallTo(whatever)` twice:

```C#
A.CallTo(() => fake.Foo()).Returns(42);
...
A.CallTo(() => fake.Foo()).MustHaveHappened();
```

It's a bit tedious, and it can also be a source of bugs. It would be nice to be able to do this instead:

```C#
A.CallTo(() => fake.Foo()).Returns(42).MustHappen();
...
A.Verify(fake);
```

**Comment by blairconrad** [comment]
Thanks for the idea, @thomaslevesque. It's interesting. I'm not entirely sold on it yet, but I never am at the beginning.

In the past, we've agonized over syntax, worrying about calls that don't "read" properly, so `A.Verify` may need work, but that's no reason not to investigate the feasibility of something like this.

None of these points are intended to kill this issue, but we'll need to consider a few things:

- one of the prime selling points of the repeated `A.CallTo` is that one doesn't have to learn a different syntax to perform verifications. This would be a new syntax, but not much more so than tacking `MustHaveHappened` onto the end of a `A.CallTo` to peform verification, and having this doesn't preclude using "classic verifications"
- `MustHaveHappened` supports a rich syntax for example minimum and maximum counts. I think if this proposal were to be pursued, it would need to support everything that `MustHaveHappened` does
- likewise, I'd think that `MustHappen` would have to be available at the end of pretty much any `A.CallTo(…)` chain


Oh. You say 

>  it can also be a source of bugs

I assume you mean the innards of the two `A.CallTo`s may not be repeated properly, so the test doesn't verify that the same calls was made as had been configured?
That problem can be ameliorated via variables:

```C#
var aCallTofakeFoo = A.CallTo(() => fake.Foo());
aCallToFakeFoo.Returns(42);
…
aCallToFakeFoo.MustHaveHappened();
```

I'm not saying it's perfect, or a completely viable alternative to your proposal, just that it's _an_ option that can make a difference in the meantime.

**Comment by thomaslevesque** [comment]
> MustHaveHappened supports a rich syntax for example minimum and maximum counts. I think if this proposal were to be pursued, it would need to support everything that MustHaveHappened does

Yes, it's also what I had in mind. I didn't put it in my example to keep things simple, but obviously `MustHappen` should accept the same arguments as `MustHaveHappened` (`Repeated.Once` and so on).

> I assume you mean the innards of the two A.CallTos may not be repeated properly, so the test doesn't verify that the same calls was made as had been configured?

Yes, that's what I mean. Any time you duplicate some code that's supposed to be the same, there's a risk of modifying it in one place and forgetting to modify it in the other place... It goes against the DRY/SPOT principles.

> That problem can be ameliorated via variables:

I thought about doing that, but it doesn't read as well as the usual syntax.


The expectation/verification model also has another benefit: even if you have set multiple expectations on a fake, you can verify them in a single step. Using `A.CallTo(...).MustHaveHappened()`, you have to check each call separately.

**Comment by thomaslevesque** [comment]
I agree that `A.Verify()` isn't great; perhaps something like `All.CallsTo(fake).HaveHappened()` would be better ?

**Comment by adamralph** [comment]
Personally, I've not liked the expectation/verification pattern since being burnt by it in previous years when using other frameworks.

The problem for me is that is makes the tests less readable. When asserting I want to see what I'm asserting, rather than having to read back and mentally untangle my expectations from my configurations.

I also believe that if return values are being configured on fake methods and it is felt important to assert that those methods are being called, then there are anti-patterns at play, e.g. [inappropriate intimacy](http://c2.com/cgi/wiki?InappropriateIntimacy).

I rarely use `MustHaveHappened()` but when I do it's almost always on some void method since the call has no measurable effect in the SUT, but even this is rare. If the SUT is depending on a value from the fake then the implementation of the fake should be concerned with the value returned. Generally I don't think it's something that the unit test should be concerned with, otherwise the unit test is effectively punching through the encapsulation, which really only leads to [Gulliver's Travels Tests](http://dhickey.ie/post/2014/03/03/gulliver-s-travels-tests.aspx).

My opinion, FWIW.

**Comment by blairconrad** [comment]
@adamralph, I tend to agree with you on all points (and had never seen _Guillier's Travels_ Tests before - nice). For my few `MustHaveHappened` uses, I'd rather type and read the `A.CallTo` twice than have to scan back to see what it is I expected. (Especially since I've seen proposed pattern (using NMock2 and Isolator) end up with `MustHappen` equivalents in the test's setup method and then you _really_ have to search.)

However, while I'm inclined against the proposal, I think it's worth noting that if it were accepted, there's nothing preventing us from keeping the existing use of `MustHaveHappened` and friends. Users could choose which pattern they prefer.

Then again, we've not been shy about saying that FakeItEasy is an opinionated framework, and this is not a situation where users will be unable to perform a task—it's down to personal preference. I don't mind using FakeItEasy's API to steer test-writers toward a certain style of test in this case.

**Comment by blairconrad** [comment]
Another point: this is the kind of behaviour that could be added on by users who would like it, perhaps not with exactly the same syntax.  For example, I just now tinkered and wrote this test:

```C#
[Test]
public void Should_be_able_to_verify_calls()
{
    var tracker = new CallTracker();

    var fake = A.Fake<IFoo>();

    tracker.MustHappen(A.CallTo(() => fake.Foo()).Returns(3));

    Assert.That(fake.Foo(), Is.EqualTo(3));

    tracker.Verify();
}
```

It passes. And fails when you take the `Foo()` call out. The syntax isn't the best, but that's mostly (I think) because I was too lazy to write an extension method, which _could_ clean things up if you don't mind having shared state on the class implementing the extension methods.
In fact, this kind of approach may even appeal to the `Verify`-likers, as it would allow one to verify calls made to several different fakes in one fell swoop.

**Comment by thomaslevesque** [comment]
> However, while I'm inclined against the proposal, I think it's worth noting that if it were accepted, there's nothing preventing us from keeping the existing use of MustHaveHappened and friends. Users could choose which pattern they prefer.

Yes, I think it's important to offer a choice. Many people prefer the expect/verify pattern, and might turn away from FIE because it doesn't offer this possibility. And introducing this feature doesn't mean you have to use it...

**Comment by adamralph** [comment]
With any new API feature there are overheads to consider. It's another thing to support, in terms of testing, documentation, potential bugs, build time (compilation and testing) and is also a potential source of confusion for newcomers. There is always an argument for keeping an API lean and focused.

With this proposal, what I'm lacking at the moment is a compelling use case, i.e. a real life example which makes expect/verify the most natural choice and therefore I'm inclined to vote against adding it. @thomaslevesque if you can provide with me that I'd be happy to reverse my view on this, but at the moment I'm just not seeing it.

**Comment by blairconrad** [comment]
I'd like to add a little weight to the "against" side. To quote @patrik-hagne in a [StackOverflow answer that talks about some of his motivation in creatingFakeItEasy](http://stackoverflow.com/questions/4001101/are-fakes-better-than-mocks/4174495#4174495)

> FakeItEasy has no "Verifiable" or "Expectations" it has assertions however, these are always explicitly stated at the very end of a test, I believe this makes tests easier to read and understand. It also helps beginners to avoid multiple asserts (where they would set expectations on many calls or mock objects).

We've been 17 days since the last update. I'm going to close the issue. It can always be reopened if a compelling argument in favour of the issue comes to light.

**Comment by DaveCousineau** [comment]
I have implemented something similar to that `CallTracker` concept. But a problem that I have with it is that I don't seem to be able to compare the result of `CallTo` to tell if I am referring to the same call or not. (The result of `A.CallTo(() => fake.Foo())` is not equal to the result of a second call, so I cannot store information about it in a dictionary, for example.)

The main thing this prevents is being able to "overwrite" previous setups with new ones. A second `MustHappen` call will create a second assertion rather than replacing the existing one.

I don't know if there is a way to compare calls to see if they are equal. Internally FakeItEasy must be doing something to keep track of calls uniquely but I'm not sure what or if it's something we can access. (Closest I've found so far is the result of `.ToString()` on the expression.)

**Comment by thomaslevesque** [comment]
Hi @DaveCousineau,

Wow, I had forgotten about this issue! I probably opened it when I was a new FakeItEasy user, before I joined the team as a maintainer... I was coming from RhinoMocks and was used to the "expect/verify" pattern, but I guess I got used to the FakeItEasy approach. I still occasionally use this approach when I have to use Moq. I don't hate it, I guess, but I don't really miss it when I use FakeItEasy.

> The main thing this prevents is being able to "overwrite" previous setups with new ones. A second `MustHappen` call will create a second assertion rather than replacing or modifying the existing one.

I'm not sure I understand why this is a problem. Could you give a concrete example?

> Internally FakeItEasy must be doing something to keep track of calls uniquely

Actually, no, we don't do anything like that, simply because we don't need to. When you configure a call, we add a new call rule to a list. When a call is made to the fake, we compare it against each rule until we find one that matches. When "overwriting" a call configuration, we actually leave the old one in place, but since newer rules are checked first, the old one is ignored.

> Closest I've found so far is the result of `.ToString()` on the expression; (that almost but doesn't quite work because a different path to the same reference results in a different string.)

No, that wouldn't work. The expression is only part of the call specification. There can also be additional conditions. Anyway, semantically different expressions could have the same string representation, so it's not a reliable approach.

Anyway, the problem you're having seems a bit abstract to me at the moment. An example with code would probably help me understand.

(FWIW, I just did a [quick POC](https://gist.github.com/thomaslevesque/5481c9cfc12b7cd6d0f719648241007b) of what you're trying to do, and I didn't see a need to test if two calls are the same)

**Comment by DaveCousineau** [comment]
Thanks for the reply. I'm trying to be able to say something like this:
```
var tracker = new CallTracker();
var fake = A.Fake<IFoo>();

tracker.SetupOne(() => fake.Foo())
.Returns(3)
.Once();

// later
tracker.SetupOneMore(() => fake.Foo())
.Returns(6)
.Once();

Assert.That(fake.Foo(), Is.EqualTo(3));
Assert.That(fake.Foo(), Is.EqualTo(6));

tracker.Verify(); // implies: A.CallTo(() => fake.Foo()).MustHaveHappened(2, Times.Exactly);
```

But I can't "remember" anything about `A.CallTo(() => fake.Foo())`. I need a key or a hashcode or something.

**Comment by thomaslevesque** [comment]
Ah, I see. Thanks for the example.

> I need a key or a hashcode or something.

Unfortunately, you won't be able to get one. Even if comparing the expressions was enough, there's no easy way to do it.


I don't see a way to make this work. My POC has the same issue: asserting the same call twice isn't the same as asserting that the call happened twice.

What you could do is, instead of making two separate configurations for the call, use `Then` to configure the two return values:
(example using my POC; not sure what it would look like with your API)

```csharp
A.CallTo(() => fake.Foo())
 .MustHappenTwiceExactly(verifier)
 .Returns(3).Once()
 .Then
 .Returns(6).Once()
```

Of course, you still need to manually match the number of calls in the assertion with the actual number of calls configured, but it's better than nothing...

Anyway, I think it will be difficult to make the whole "expect/verify" approach work correctly with all nuances of FakeItEasy's call assertions (e.g. [ordered assertions](https://fakeiteasy.github.io/docs/7.4.0/ordered-assertions/) would probably be tricky)

**Comment by blairconrad** [comment]
> Anyway, I think it will be difficult to make the whole "expect/verify" approach work correctly with all nuances of FakeItEasy's call assertions (e.g. [ordered assertions](https://fakeiteasy.github.io/docs/7.4.0/ordered-assertions/) would probably be tricky)

Indeed. I was thinking about this last night, and there are restricted cases (no ordering, and maybe no limits on the numbers of calls made) where a "verify" approach would be possible, but FakeItEasy's arrange/assert model is too rich for everything to be covered, IMO.

I'm not trying to question the way you work, @DaveCousineau, but I mostly find no need for a "verify" step (whether written as `Verify` or `A.CallTo(…).MustHaveHappened()`). The overwhelming majority of my tests are written in such a way that if the call configurations I've arranged aren't exercised (because the call was never made or was made improperly), the assertions I make on my system under test's behaviour fail, so I omit any verification on the Fake.

I sort of feel like in most cases, the verification step is redundant (and a little bit constrictive; if I change the test later, I have to change the arrange and the verification), which I think might be what prompts you (and young @thomaslevesque, and lots of people) to want an easy way to arrange a behaviour and remember it for later verification all in one step. 

Instead, I find that in 90% or more of the test cases I write, I only need arrange. When I _do_ want a verification step (almost always when I'm making sure my system under test caches results or some such), the effort of writing the assertions on the Fake is not so bad, and I almost enjoy the fact that these tests look different than most. It's another way to reveal intention.

Anyhow, thanks for listening. Like I said, I'm not trying to question your need for the feature, but to provide a little insight on how (I, at least, think) FakeItEasy's separate "mock verification" steps often work quite well (and as noted, are probably more powerful than we'd ever be able to build into a one-stop verification method).

----

aside:

I'm curious about your API. I see

```c#
tracker.SetupOne(() => fake.Foo()).Returns(3).Once();
tracker.SetupOneMore(() => fake.Foo()).Returns(6).Once();
```

And I wonder about `SetupOne` and `SetupOneMore`. It seems like those calls are encoding something special about the order and/or number of times the methods are called, and it already a form of duplication. Maybe this is only a stepping-stone, and the goal was to have a more consistent setup phase, with `tracker` taking care of more of these details internally?

**Comment by DaveCousineau** [comment]
The idea is that I can have default setup code that is used by a bunch of related tests but can also be added onto when needed. What I've done so far is basically this:
```
public void DefaultSetup(
   int fooAdd = 0
) {
   SetupExactly(() => fake.Foo(), 1 + fooAdd)
   .Returns(3)
   .Once();
}

...

tracker.DefaultSetup(fooAdd: 1);

// this would be tracker.SetupOneMore(() => fake.Foo())
// if I could do that
A.CallTo(() => fake.Foo())
.Returns(6)
.Once();

// works
Assert.That(fake.Foo(), Is.EqualTo(3));
Assert.That(fake.Foo(), Is.EqualTo(6));

tracker.Verify();
```
It's a bit clumsy and I have to make sure to use `A.CallTo` and not call one of my Setup methods the second time, but it works.

> The overwhelming majority of my tests are written in such a way that if the call configurations I've arranged aren't exercised (because the call was never made or was made improperly), the assertions I make on my system under test's behaviour fail, so there's no downside to omitting verification on the Fake.

You're probably right. My situation is that I have only just started using mocking frameworks a few weeks ago and I'm still working out how best to use them. I was using Moq at first but a certain incident ended up ocurring and I decided to switch to something else and the two candidates were NSubstitute and FakeItEasy. NSubstitute seems not support strict mocks, so I went with FakeItEasy. The `Verify()` thing I had gotten used to with Moq. So far what I do is write things as strictly as possible until it seems clear that something should be loosened.

One example I have for the verify approach is `Dispose()`. Many of my tests look like this:
```
TestingData td;
using (td = new()) {
   ...
}

td.Verify();
```
And I have methods that automatically hook up disposables (or other methods that assert that something is not disposable), and then the `Verify()` call will assert that anything that was disposable did actually have `Dispose()` called on it. The alternative would be to have a disposing test with a bunch of `A.CallTo(() => x.Dispose()).MustHaveHappenedOnceExactly();` calls for every mock that is disposable. It's a bit nicer that anything that I add as a disposable just automatically gets verified as having been disposed.

I may likely change my mind about some things as I get more used to writing these kinds of tests.
